### ISSUE TITLE
Research and Provide a Demo of Unit Test Implementation for HfLA Repo

### ISSUE THREAD

**Issue by adam-abundis** [hackforla/website#2249]
### Overview
We need to see if it is feasible to implement unit tests for JavaScript functionality within the Hack for LA website repo, both within GitHub Action files and within client-side JavaScript.  Ideally the same test framework would be used for unit testing both the back end and front end JavaScript.  

### Action Items
- [x] Research implementation of unit tests in a Jekyll environment, with the goal of finding a test framework that could be used for automated testing of both the back end (GitHub Actions) and front end (client-side) JavaScript in the HfLA website repo.  For the purpose of this issue, the test can be executed manually from the command-line, but it must be possible to incorporate into an automation in a future issue
   - [x] Suggestions: Mocha, Jasmine
- [ ] Create a new branch and implement a library that provides unit testing
   - [ ] Add unit test for 1 JavaScript file within the HfLA website repo
   - [ ] Add unit test for 1 Github action file
- [ ] Create Draft Pull request with the unit test implementation for the dev team to review 
- [ ] Demo the unit test implementation for the back-end meeting on Thursday
- [ ]  Determine with the rest of the dev team whether adding unit tests is feasible after the demo   

### Resources/Instructions
https://mochajs.org/
https://jasmine.github.io/
[Karma test runner](https://karma-runner.github.io/3.0/)

**Comment by arpitapandya** [comment]
Availability for this week: 6hrs
My estimated ETA for completing this issue: 10hrs

**Comment by arpitapandya** [comment]
Availability : 6hrs
Progress: I am working on it.
Blockers: None
ETA: 2022-09-19

**Comment by arpitapandya** [comment]
I have unassigned in order to gain more understanding to create GHA first before I test

**Comment by ExperimentsInHonesty** [comment]
@hackforla/website-merge not sure what should be done with the items that are checked.  Please take a look

**Comment by nnr-nnr** [comment]
Availability:  3–8 PM M–F
ETA: 9 PM Thurs. 3/30/23

**Comment by nnr-nnr** [comment]
Progress Update 
- There's seems to be limited info on testing in a Jekyll environment, so I think i'm going to stick with the linked resource.
- I got some questions answered at OHs about the logistics, ie that there is no backend team meeting and I could just present at a normal dev meeting. Also, that the "Draft Pull request" can just be treated as a PR that won't be merged.
- No blockers.
- Will work through the checklist this weekend.

**Comment by nnr-nnr** [comment]
> Research implementation of unit tests in a Jekyll environment. Provide the results of the research in a comment within this issue.


There are limited information and resources on the web about unit testing in a Jekyll environment. Besides the library attached in this Issue ([Joule](https://github.com/helpscout/jekyll-joule), last updated 2017), I found [suggestions](https://talk.jekyllrb.com/t/advice-for-including-javascript-unit-tests/4398) of using GitHub actions to test code before pushing. [Others](https://medium.com/espark-engineering-blog/testing-your-jekyll-site-the-ruby-way-ccfa386a8318) suggest using the following together for a complete acceptance testing framework for Jekyll sites:

- [RSpec](http://rspec.info/) — Behavior Driven Development for Ruby
- [Capybara](https://github.com/jnicklas/capybara) — Testing of web applications via simulation of a real user’s interaction with your app
- [Selenium](http://www.seleniumhq.org/) — Browser automation

**Comment by nnr-nnr** [comment]
BLOCKER:
I cannot do unit testing with [Joule](https://github.com/helpscout/jekyll-joule) because I do not have permissions to install Joule in the repo. Running this command in terminal:
`gem install jekyll-joule`
resulted in this error message:
```
Fetching addressable-2.8.3.gem
Fetching sass-embedded-1.61.0-x86_64-darwin.gem
ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /Library/Ruby/Gems/2.6.0 directory.
```

**Comment by nnr-nnr** [comment]
I'm running into issues with the [setup tasks](https://github.com/helpscout/jekyll-joule/blob/master/docs/setup.md) of the Joule testing library. The setup requires creating a Gemfile and Rakefile. I was unable to get the site up locally with Docker after applying these changes. I'm not sure if these changes are compatible with the structure of the current website repo. Any resources discussing how to integrate a new library with the current repo, dependencies, etc. would be helpful to start off.

** [Joule](https://github.com/helpscout/jekyll-joule) notes: "So far, Joule has only been tested with [minitest](https://github.com/seattlerb/minitest). However, you should be able to use it with other testing frameworks."

**Comment by ExperimentsInHonesty** [comment]
@nnr-nnr I saw that you put a comment https://github.com/hackforla/website/issues/2249#issuecomment-1522664094, asking for resources but didn't at mention the merge team.  If you have not received help, I recommend posting the comment link into the slack channel and see if anyone can lend a hand.

**Comment by ExperimentsInHonesty** [comment]
Will wait for assignee to respond (1 week), if no response, remove from issue and reclassify as Large.

**Comment by nnr-nnr** [comment]
Thank you @ExperimentsInHonesty. I'll share the link to this comment explaining my blocker in Slack.
——

I am trying to implement [Joule](https://github.com/helpscout/jekyll-joule), a Jekyll utility helper for writing unit tests, into the hackforla/website repo. I have followed the library’s [setup instructions](https://github.com/helpscout/jekyll-joule/blob/master/docs/setup.md). In particular, I have:

1. [Installed Joule](https://github.com/helpscout/jekyll-joule#install) in the repo with this commandline code: `sudo gem install jekyll-joule`. This required me to install other dependencies, including `sass-embedded -v 1.58.3` and `rouge -v 3.30.0`.
2. [Added a Gemfile](https://github.com/helpscout/jekyll-joule/blob/master/docs/setup.md#add-test-dependencies) to the root directory (such a file didn’t previously exist) with the following:
```
source 'https://rubygems.org'
group :test do
  gem "minitest", "~> 5.8"
  gem "minitest-profile"
  gem "minitest-reporters"
  gem "rake", "~> 10.0"
  gem "rspec-mocks"
  gem "shoulda"
  gem "jekyll-joule"
end
```
Follow up Q: if not  with a Gemfile, how are dependencies/Jekyll libraries manages in the website repo?
3. [Added a Rakefile](https://github.com/helpscout/jekyll-joule/blob/master/docs/setup.md#add-a-rakefile) in the root directory with the following:
```
require 'rake'
require 'rake/testtask'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), *%w[lib]))

Rake::TestTask.new(:test) do |test|
  test.libs << 'lib' << 'test'
  test.pattern = 'joule-tests/**/test_*.rb'
  test.verbose = true
end
```
 Note: a “test” folder already exists in the root directory, so I created one named `joule-tests` (see below).

4. [Added a test directory](https://github.com/helpscout/jekyll-joule/blob/master/docs/setup.md#add-a-test-directory) in the root folder named `joule-tests`.
6. [Added a helper.rb file](https://github.com/helpscout/jekyll-joule/blob/master/docs/setup.md#add-a-test-helper-file) to the root directory with [this](https://github.com/helpscout/jekyll-joule/blob/master/examples/test/helper.rb) code.
7. [Added a test file](https://github.com/helpscout/jekyll-joule/blob/master/docs/writing-tests.md#add-your-test-file) named `test_yiss.rb`  to the `joule-tests` directory with [this](https://github.com/helpscout/jekyll-joule/blob/master/docs/writing-tests.md#write-your-test) code.
8. [Ran the test](https://github.com/helpscout/jekyll-joule/blob/master/docs/writing-tests.md#run-your-test) with `bundle exec rake test` on the commandline.
9. Trouble shooting a bit, I moved the `helper.rb `code directly into the test_yiss.rb.
10. I'm now running into two major issues:
a) `bundle exec rake test` produces the following output:
> /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby -I"lib:lib:test" -I"/Library/Ruby/Gems/2.6.0/gems/rake-10.5.0/lib" "/Library/Ruby/Gems/2.6.0/gems/rake-10.5.0/lib/rake/rake_test_loader.rb" "joule-tests/**/test_*.rb" 
> Started with run options --seed 41668
> 
> AwwYissTest
>   test:AwwYiss should render a div.                              ERROR (0.01s)
> Minitest::UnexpectedError:         NoMethodError: undefined method `fix_common_issues' for #<Jekyll::Configuration:0x00007fd45891d348>
>             /Users/nivarazin/Desktop/Projects/hackforla/website/joule-tests/test_yiss.rb:58:in `build_configs'
>             /Users/nivarazin/Desktop/Projects/hackforla/website/joule-tests/test_yiss.rb:62:in `site_configuration'
>             /Users/nivarazin/Desktop/Projects/hackforla/website/joule-tests/test_yiss.rb:27:in `setup'
> 
> 
> Finished in 0.01082s
> 1 tests, 0 assertions, 0 failures, 1 errors, 0 skips
> rake aborted!
> Command failed with status (1): [ruby -I"lib:lib:test" -I"/Library/Ruby/Gems/2.6.0/gems/rake-10.5.0/lib" "/Library/Ruby/Gems/2.6.0/gems/rake-10.5.0/lib/rake/rake_test_loader.rb" "joule-tests/**/test_*.rb" ]
> /Library/Ruby/Gems/2.6.0/gems/bundler-1.17.2/exe/bundle:30:in `block in <top (required)>'
> /Library/Ruby/Gems/2.6.0/gems/bundler-1.17.2/exe/bundle:22:in `<top (required)>'
> /usr/bin/bundle:23:in `load'
> /usr/bin/bundle:23:in `<main>'
> Tasks: TOP => test
> (See full trace by running task with --trace)

I'm not sure the cause of the error, but I wonder if it has to do with other dependency versions in the repo based on Googling. Also, it looks like the Joule library hasn't been updated in >5 years...

b) At this point, running `docker-compose up` produces the following error-riddled output:

> [+] Running 1/0
>  ⠿ Container hfla_site  Created                                                                                            0.0s
> Attaching to hfla_site
> hfla_site  | touch: /connected: Permission denied
> hfla_site  | ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux-musl]
> hfla_site  | /usr/local/lib/ruby/2.7.0/rubygems/dependency.rb:313:in `to_specs': Could not find 'bundler' (1.17.2) required by your /srv/jekyll/Gemfile.lock. (Gem::MissingSpecVersionError)
> hfla_site  | To update to the latest version installed on your system, run `bundle update --bundler`.
> hfla_site  | To install the missing version, run `gem install bundler:1.17.2`
> hfla_site  | Checked in 'GEM_PATH=/home/jekyll/.gem/ruby/2.7.0:/usr/local/lib/ruby/gems/2.7.0:/usr/gem', execute `gem env` for more information
> hfla_site  | 	from /usr/local/lib/ruby/2.7.0/rubygems/dependency.rb:323:in `to_spec'
> hfla_site  | 	from /usr/local/lib/ruby/2.7.0/rubygems/core_ext/kernel_gem.rb:62:in `gem'
> hfla_site  | 	from /usr/local/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:71:in `require'
> hfla_site  | 	from /usr/gem/gems/jekyll-3.9.2/lib/jekyll/plugin_manager.rb:48:in `require_from_bundler'
> hfla_site  | 	from /usr/gem/gems/jekyll-3.9.2/exe/jekyll:11:in `<top (required)>'
> hfla_site  | 	from /usr/gem/bin/jekyll:25:in `load'
> hfla_site  | 	from /usr/gem/bin/jekyll:25:in `<main>'
> hfla_site exited with code 1

Running `bundle update --bundler` and `gem install bundler:1.17.2` has not resolved these error messages.
Because I'm just doing unit testing, I don't think it's super necessary for me to view the website/have a docker container running. However, it's obviously not good that this is throwing an error.

———
Actually writing tests with Joule seems [somewhat straightforward](https://github.com/helpscout/jekyll-joule#basic-usage), as it just checks the existence and properties of objects in the DOM. The setup is proving tricky, though. Any ideas/suggestions are welcome!

**Comment by roslynwythe** [comment]
Hi @nnr-nnr please leave an update.  I understand that you will be away for a bit and wanted to check if you wanted us to unassign you from this issue.

**Comment by nnr-nnr** [comment]
Other HfLA members shared with me that [Selenium](https://www.selenium.dev/) is an up-to-date, open-source browser testing tool that may fit this use case well. I'm really interested in doing more research into Selenium, particularly because Jekyll Joule hasn't been updated in >5 years and is much less widely used. I misconfigured my dev environment while trying to setup Jekyll Joule. I installed some dependencies that would not allow Docker to run. I tried to uninstall them but that didn't help. So I'm a bit stuck here and switching branches doesn't help my environment, so I can't review other branches, either. When I'm freed up again (in ~1 week), I'll seek help about fixing my environment and share research about Selenium. I'm not well-versed in DevOps, so that may be the extent that I can do for this issue.

**Comment by roslynwythe** [comment]
Hi Niva, Sarah suggested that you delete the docker images and then rebuild the website locally.  If that doesn't work, you could wipe out the repo and and delete the docker images and re-setup the repo.  If you can attend Thursday's office hours, she would be happy to work with you.

**Comment by roslynwythe** [comment]
Hi @nnr-nnr, please leave us an update on this issue.  Thanks!

**Comment by nnr-nnr** [comment]
This issue requires more DevOps background than I have. The instructions suggest Joule as a testing framework, but that library hasn't been updated in many years. Someone else suggested [Selenium](https://www.selenium.dev/), which may be a better starting point for the next contributor to this issue. Also, I think it was decided that this issue should be changed to Complexity:Large.

**Comment by ExperimentsInHonesty** [comment]
@nnr-nnr Thank you for the notes.  I have moved this issue back to the new issue approval column, so that a dev lead can look at them and revise this issue. 

-Bonnie

**Comment by roslynwythe** [comment]
- I removed reference to Joule and added suggestion to research Mocha and Jasmine, and emphasized the desire to select a framework that could be used for unit testing of both JavaScript in GHA as well as JavaScript in client-side JavaScript.  This is a challenging requirement and so I increased the complexity to Large.  

- In comments above, the suggestion was made to consider Selenium.  I did not mention Selenium in my edits to the issue because it seems like an end-to-end test rather than unit test approach, and it is not appropriate for testing JavaScript in GHAs,  but I like the idea and so I created #6017

**Comment by danvgar** [comment]
Happy to take up this issue, this sounds really interesting. I don't have a strong idea of an ETA yet but I will look into this by EOW Sun Mar 24 and provide an update on how long I think this should take. Typically available 2-7pm PST every day if anyone needs to reach me.

3/26 Edit: I am still reviewing this issue and will have an update shortly. Thank you!

**Comment by danvgar** [comment]
Apologies for the delays, the past two weeks have been a bit hectic in my personal life. I am currently working on these today and tomorrow (Sat Apr 6) and will have an update by EOD Sun Apr 7

**Comment by danvgar** [comment]
@roslynwythe @bonniewolfe Sincerest apologies, I've had very low bandwidth the past few weeks. This next week is still tricky but I will do my best this weekend to tackle the updates you provided over the past week. Thank you!

**Comment by danvgar** [comment]
Unassigning myself as I haven't been able to dedicate time in the past weeks to this issue. Hopefully someone else can take it up but if not I will re-assign it to myself when I have more bandwidth! Apologies for the delays

**Comment by Dartis4** [comment]
Availability: 10am - 5pm
ETA: Undetermined as of right now

**Comment by Dartis4** [comment]
I think the "testing in a Jekyll environment" may have been a red herring to this whole issue. To my understanding, Jekyll is just a static site generator, it should not require any testing, it is everything else that would require the unit tests. From my research thus far, it seems like being able to do unit testing for GitHub Actions and the frontend javascript code is pretty straight forward.

According to [this blog article](https://dev.to/cicirello/how-to-test-a-github-action-with-github-actions-2hag), you can test GitHub Actions from within a GitHub Actions Workflow using premade data/website files for testing. He has a few examples where he simply creates a GitHub Action that runs the unit tests for his code (Python code in this case) along with integration testing. In the Actions file you can create extra run steps that will execute all of the preliminary build and execution and then dictate which specific data should be used for that particular run of the Action. This essentially mocks what changes are the target of the action, and if any of the unit testing fails on that run iteration of the Action, the workflow will fail and the process stops. 

This method he presents would be effective, however cumbersome as it would require pushing to the remote branch and waiting for the test workflow to run in order to test new changes. The comments of the blog post gave some suggestions for work arounds to this issue. The first was [this](https://github.com/cardinalby/github-action-ts-run-api), which enables running GitHub and Docker Actions locally rather than directly in GitHub. However, the library is implemented in entirely TypeScript, which it appears we do not use. This would add complexity to the implementation of this library. he [second suggestion](https://github.com/nektos/act) could overcome this issue as it runs the whole workflow locally using existing workflows! This would effectively reduce the feedback loop of testing GitHub Actions by allowing them to be run locally from the commandline (and potentially in a GitHub Action to automate the process), and simplify the implementation by using existing workflows rather than requiring the creation of new workflows and actions.

In terms of Javascript testing, it seems like Mocha and Jasmine are great options for testing frameworks. Jasmine appears to be a more straight forward and simple framework that would allow for quick implementation and is more fully featured out of the box with cleaner documentation. Mocha has messier documentation, but might have more niche testing features for certain use cases. I would probably recommend Jasmine.

**To summarize**, we can use the `act` library from [nektos](https://github.com/nektos/act) (from the second suggestion above) to run local GitHub Actions that will test our exisiting workflows and perform both unit testing for our GitHub Actions and our JavaScript code locally. Jasmine is likely an appropriate testing framework for our purposes.

I will look into creating examples of this working for our repo. @roslynwythe I would love your thoughts on this idea in the mean time and whether you think this is a viable route to go down.

**Comment by roslynwythe** [comment]
@Dartis4 First, unit testing the front-end JavaScript would appear to present certain challenges, because the code (as it is stored in GitHub) contains YAML/liquid statements that must be processed by Jekyll to produce pure JS, then in the browser it is transpiled by Babel.js and finally it runs in the context of a browser/server connection.  But hopefully the test runner can setup the proper environment.  

Regarding testing of GitHub actions, there was some use of nektos act in the past, but currently our backend devs seem to prefer testing GHAs in their GitHub repositories.  Typically during testing, the GHA have workflow dispatch enabled so that workflows can be manually triggered, and the developer has a copy of the HfLA project board and the required secrets to test the GHA fully.   I hesitate to head in a different direction without discussion with the team, especially because our goal with unit testing is to integrate tests as GitHub checks in the context of Pull Requests.  

To begin discussion, I will create a Slack post inviting comment on the issue.   Thank you for great work so far on this issue.

**Comment by Dartis4** [comment]
Progress: Found a few leads for testing frameworks. Need to try implementing into the work flow to see ig things are viable. In talks with Roslyn and Will about viability of the libraries as well. 
Blockers: None at the moment.
Availability: I should be available between 10am - 5pm each day this week.
ETA: Unknown

**Comment by Dartis4** [comment]
Progress: I believe this issue should be split into two separate issues and close this issue. The two new issues should be one focused on GitHub Actions unit testing, the other on frontend unit testing.
Blockers: None
Availability: 10am - 5pm
ETA: End of this week

**Comment by roslynwythe** [comment]
- @Dartis4 please see #7151 .  Note the `Draft` label.  Your comments would be appreciated.  You are welcome to edit the issue directly or provide suggestions to me and I will edit.  If you wish to be assigned to the issue, please create a comment, requesting the assignment from @ExperimentsInHonesty.  Thank you for your excellent research and analysis.

**Comment by roslynwythe** [comment]
- @ExperimentsInHonesty I would like to close this issue and replace it with #7151 and #7152
